File: apps/desktop/src-tauri/src/quest.rs
use std::io::Result;

use editor::{quest::quest_file::QuestFile, questlist::quest_info::QuestInfo};
use serde::{Deserialize, Serialize};

use crate::utils::{wrap_json_result, wrap_result};

#[derive(Serialize, Deserialize, Debug, PartialEq)]
struct SaveQuestPayload {
    filepath: String,
    quest: QuestFile,
}

#[derive(Serialize, Deserialize, Debug, PartialEq)]
struct ExportQuestInfoPayload {
    filepath: String,
    quest_info: QuestInfo,
}

#[tauri::command]
pub fn save_quest_file(event: String) -> String {
    let result = || -> Result<String> {
        let mut payload = serde_json::from_str::<SaveQuestPayload>(&event)?;
        QuestFile::save_to(&payload.filepath, &mut payload.quest)?;

        Ok(String::from("{ \"status\": \"Success\" }"))
    };

    match result() {
        Ok(response) => response,
        Err(error) => wrap_result(error.to_string(), true),
    }
}

#[tauri::command]
pub fn read_quest_file(event: String) -> String {
    let result = QuestFile::from_path(&event);
    wrap_json_result(result)
}

#[tauri::command]
pub fn export_quest_info(event: String) -> String {
    let result = || -> Result<String> {
        let mut payload = serde_json::from_str::<ExportQuestInfoPayload>(&event)?;
        payload.quest_info.save_to(&payload.filepath)?;

        Ok(String::from("{ \"status\": \"Success\" }"))
    };

    match result() {
        Ok(response) => response,
        Err(error) => wrap_result(error.to_string(), true),
    }
}

File: src/events/quest.ts
import { prepareQuest } from "../utils/prepare";
import { QuestFile, QuestInfo } from "../utils";
import { invokeEvent } from "./core";

interface ExportQuestInfoPayload {
  filepath: string,
  quest_info: QuestInfo
}

export const exportQuestInfo = async (payload: ExportQuestInfoPayload): Promise<void> => {
  await invokeEvent<ExportQuestInfoPayload, void>(
    'export_quest_info',
    payload
  );
}

export const readQuest = async (path: string): Promise<QuestFile> => {
  return invokeEvent<string, QuestFile>(
    'read_quest_file',
    path
  );
};

export interface SaveQuestPayload {
  filepath: string;
  quest: QuestFile;
}

export const saveQuest = async (data: SaveQuestPayload) => {
  const quest: QuestFile = prepareQuest(data.quest);
  const payload: SaveQuestPayload = { filepath: data.filepath, quest };

  await invokeEvent(
    'save_quest_file',
    payload
  );
};

export const readQuestInfo = async (path: string): Promise<QuestInfo> => {
  return invokeEvent<string, QuestInfo>(
    'read_questinfo',
    path
  );
}